mitodl/mitxpro
# feat: dollars off discount support in basket/checkout

#### Pre-Flight checklist


#### What are the relevant tickets?
#2448

#### What's this PR do?
- Adds support for dollars-off coupons on the basket
- Refactors multiple checkout pages to support dollars-off discounts
- Adds dollars-off support for HubSpot
- Refactors and adds tests

#### How should this be manually tested?
This might need to be tested extensively. It might be a good idea to read the conversation on https://github.com/mitodl/mitxpro/issues/2443 before starting the testing.
- Create a couple of coupons 'dollars-off' or 'percent-off; either from Django Admin or through `http://xpro.odl.local:8053/admin/`
- Apply those coupons on an item in the basket (Make sure that all the cases are tested e.g. Global or course-based coupons work exactly as they should for both `dollars-off` and `percent-off`)
- Create dollars-off and percent-off coupons globally and test the case where we pick the best coupon automatically
- Check that the discounts and price displayed to the user is correct
- Disable coupons and check if they become invalid for both `dollars-off` and `percent-off`
- Check the Hubspot that the orders are created successfully there and also they contain the right discount and amounts as mentioned in https://github.com/mitodl/mitxpro/issues/2443
- For Hubspot test keys (You can ask me)
- Also check that the discount or price values never show up to be `negative` since we have dollars-off support now and with dollars, the values can be negative
- No `dollars-off` support for B2b coupons are only percent based and work as expacted and coupons generated through b2b have `discount_type` to be `percent-off`
- No dollars-off support for coupons through sheets so they should work as expacted and have `discount_type` to be `percent-off`.

#### Where should the reviewer start?
Checkout

#### Any background context you want to provide?
- This is part of the epic and for that, the related work is also done in https://github.com/mitodl/mitxpro/pull/2449 and https://github.com/mitodl/mitxpro/pull/2451
